gnome-initial-setup/gnome-initial-setup.c

```diff
@@ -156,6 +158,81 @@ destroy_pages_after (GisAssistant *assistant,
   }
 }
 
+static gint
+compare_pages_order (gconstpointer a,
+                     gconstpointer b,
+                     gpointer data)
+{
+  GHashTable *sort_table = data;
+  const gchar *id_a = ((PageData *) a)->page_id;
+  const gchar *id_b = ((PageData *) b)->page_id;
+  gint order_a = (gint) GPOINTER_TO_UINT (g_hash_table_lookup (sort_table, id_a));
+  gint order_b = (gint) GPOINTER_TO_UINT (g_hash_table_lookup (sort_table, id_b));
+
+  return order_a - order_b;
+}
+
+static void
+reorder_pages (GisDriver *driver)
+{
+  gint num_pages;
+  GKeyFile *key_file = gis_driver_get_vendor_conf_file (driver);
+  g_auto(GStrv) conf_pages_order = NULL;
+  g_autoptr(GHashTable) sort_table = NULL; /* (owned) (element-type utf8 uint) */
+  g_autoptr(GError) error = NULL;
+
+  if (key_file == NULL)
+    return;
+
+  conf_pages_order = g_key_file_get_string_list (key_file, "pages", "order",
+                                                 NULL, &error);
+  if (conf_pages_order == NULL) {
+    /* if we have the pages order key but there was an issue getting it */
+    if (!g_error_matches (error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_KEY_NOT_FOUND) &&
+        !g_error_matches (error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_GROUP_NOT_FOUND))
+      g_debug ("Couldn't get pages order from the conf file: %s", error->message);
+
+    return;
+  }
+
+  if (*conf_pages_order == NULL) {
+    g_debug ("No pages defined in the pages order from the conf file");
+    return;
+  }
+
+  /* assign the same sort weight to every page (this will make sure) that any
+   * page to be sorted differently will show up before pages that aren't in the
+   * custom sorting list */
+  sort_table = g_hash_table_new (g_str_hash, g_str_equal);
+
+  for (num_pages = 0; page_table[num_pages].page_id != NULL; ++num_pages) {
+    g_hash_table_insert (sort_table, (gchar *) page_table[num_pages].page_id,
+                         GUINT_TO_POINTER (PAGES_DEFAULT_SORT_WEIGHT));
+  }
+
+  /* assign the new sort weights from the conf file */
+  for (guint i = 0; conf_pages_order[i] != NULL; ++i) {
+    gchar *page_id = conf_pages_order[i];
+    if (g_hash_table_lookup (sort_table, page_id) == NULL) {
+      g_debug ("No page with id '%s' found while sorting the pages!", page_id);
+      continue;
+    }
+
+    g_hash_table_insert (sort_table, page_id, GUINT_TO_POINTER (i));
+  }
+
+  g_qsort_with_data (page_table,
```

Comment:
I had to check the documentation to see if this is a stable sort (ie, preserves the existing order of pages which compare equal according to `compare_pages_order`). It is, since GLib 2.32, but please add a comment to this effect.

Reply:
I know it's stable and I mentioned the order being kept in the commit message. Also, g-i-s requires glib 2.53.0.
In any case, I don't think a comment makes a lot of difference here. If we need to make sure this doesn't ship if suddenly g-i-s is changed to use glib < 2.32 (very unlikely), we should use some `ifdefs` or whatever with the glib version.

Reply:
I'm not worried that it will break; and of course not worried that we will suddenly downgrade to a prehistoric GLib. I was worried because I know that Quicksort is, in general, not stable. `man 3 qsort` says:

> If two members compare as equal, their order in the sorted array is undefined.

So upon reading the code, it was not clear to me that the claim in the commit message:

> […] rest of the pages, in the default order they had among themselves

was true. This is why I suggested a comment. If you think it is obvious that `g_qsort_with_data` preserves order for equal elements, okay.